tsl/src/compression/compression.c

```diff
@@ -1450,3 +1452,58 @@ compression_get_toast_storage(CompressionAlgorithms algorithm)
 		elog(ERROR, "invalid compression algorithm %d", algorithm);
 	return definitions[algorithm].compressed_data_storage;
 }
+
+/* Get relstats from compressed chunk and insert into relstats for the
+ * corresponding uncompressed parent.
+ */
+extern void
+update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
+{
+	bool update = true;
+	double rowcnt;
+	int comp_pages, uncomp_pages, comp_visible, uncomp_visible;
+	float comp_tuples, uncomp_tuples, out_tuples;
+	Chunk *uncompressed_chunk = ts_chunk_get_by_relid(uncompressed_relid, true);
+	Chunk *compressed_chunk = ts_chunk_get_by_relid(compressed_relid, true);
+
+	if (uncompressed_chunk->table_id != uncompressed_relid ||
+		uncompressed_chunk->fd.compressed_chunk_id != compressed_chunk->fd.id ||
+		compressed_chunk->table_id != compressed_relid)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("mismatched chunks for relstats update %d %d",
+						uncompressed_relid,
+						compressed_relid)));
+	}
+
+	/* we need page info from compressed relid */
+	capture_pgclass_stats(compressed_relid, &comp_pages, &comp_visible, &comp_tuples);
+	rowcnt = (double) ts_get_before_compression_row_count(uncompressed_chunk->fd.id);
+	capture_pgclass_stats(uncompressed_relid, &uncomp_pages, &uncomp_visible, &uncomp_tuples);
+
+	/* Before compressing a chunk in 2.0, we save its stats. Prior
+	 * releases do not support this. So the stats on uncompressed relid
+	 * could be invalid. In this case, do the best that we can.
+	 * Alternately, the stats could match, don't update in that case.
+	 */
+	if (uncomp_tuples == 0)
+	{
+		if (rowcnt > 0)
+			out_tuples = (float4) rowcnt;
+		else
+			out_tuples = (float4) comp_tuples;
+	}
+	else
+	{
+		out_tuples = (float4) uncomp_tuples;
+		/* tuple count is unchanged. If num pages changed, we will update */
+		if (comp_pages == uncomp_pages)
```

Comment:
This doesn't seem right to me, as we'd expect comp_pages to be less than uncomp_pages here.  I'm not sure how this information is used in the planning, but given that it's passed up to the parent, I'd assume it's more correct to return the number of pages expected to be passed up by this child (so uncomp_pages).

Assuming I've convinced you that uncomp_pages is more correct here, I also think that if we're updating the number of rows (rowcnt != uncomp_tuples) then we should update the number of tuples to try to preserve the same tuple:page ratio.

Reply:
It would be equal on 2.0 due to the compress_chunk behavior that updates stats in 2.0? But we could vacuum the compressed hypertable and find that we now have fewer pages. That's the reason for the update. uncomp-apges might be 0 if it was vacuumed. Or it would be == comp_pages due to compress_chunk?